### PR TITLE
Gaia astroquery 1.0

### PR DESCRIPTION
**Pull request by mhsarmiento**
this new pull request contain a new enhancement for the packages 'utils.tap' and 'Gaia.'  This new enhancements includes the capability to rename a user table and/or its columns.

### CODE REVIEW

File: CHANGES.rst
Comment (keflavich) [inline CHANGES.rst:12]: style issue: should be
```suggestion
utils.TAP
^^^^^^^^^
```

File: CHANGES.rst
Comment (keflavich) [inline CHANGES.rst:17]:
```suggestion
gaia
^^^^
```

File: astroquery/gaia/core.py
Comment (keflavich) [inline astroquery/gaia/core.py:917]: the docstring is overindented.

File: astroquery/gaia/core.py
Comment (keflavich) [inline astroquery/gaia/core.py:907]: If this is meant to be an example for users to follow, it belongs in the
```
Example
-------
```
section below `Parameters`.

File: astroquery/gaia/core.py
Comment (keflavich) [inline astroquery/gaia/core.py:905]: Minor English correction
```suggestion
            This new method allows you to update the column names of a user table.
```

File: astroquery/utils/tap/core.py
Comment (keflavich) [inline astroquery/utils/tap/core.py:1698]: remove these unnecessary comments

**Review by keflavich — CHANGES_REQUESTED** [review]
I have a few stylistic change requests, but otherwise I think this is in good shape.

**Comment by bsipocz** [comment]
@mhsarmiento - you could please also rebase this and address the review?

File: docs/gaia/gaia.rst
Comment (bsipocz) [inline docs/gaia/gaia.rst:951]: minor typo
```suggestion
    >>> gaia.login()
```

File: docs/gaia/gaia.rst
Comment (bsipocz) [inline docs/gaia/gaia.rst:950]: what is `<env>`? I don't see mentioned anywhere above (unlike `user_<user_login_name>`). I think an example would be useful.

**Review by bsipocz — COMMENTED** [review]
It appears that I had a review started a while ago but it hasn't been submitted yet, I'm sorry for the delay.

There are a couple of comments, most of them are minor.

File: astroquery/utils/tap/core.py
Comment (eerovaher) [inline astroquery/utils/tap/core.py:1708]: Shouldn't this only be printed if the user has set `verbose` to be `True`?

File: astroquery/utils/tap/core.py
Comment (eerovaher) [inline astroquery/utils/tap/core.py:1725]: Instead of checking if `key == ''` wouldn't it be better to check if `key` is actually present in the table? Or is that not possible?

File: astroquery/utils/tap/core.py
Comment (eerovaher) [inline astroquery/utils/tap/core.py:1728]: Wouldn't it be better to check if the new column name is already in the table or not? Or is that not possible?

File: astroquery/utils/tap/core.py
Comment (eerovaher) [inline astroquery/utils/tap/core.py:1730]:
```suggestion
                raise ValueError(f"New name of {key} cannot be an empty string")

```

File: astroquery/utils/tap/core.py
Comment (eerovaher) [inline astroquery/utils/tap/core.py:1737]: Assuming that the preceding code has verified all the data then the following would be more pythonic:
```suggestion
        new_column_names = ','.join(f'{key}:{value}' for key, value in new_column_names_dict.items())
```

File: astroquery/utils/tap/core.py
Comment (eerovaher) [inline astroquery/utils/tap/core.py:1748]: The only difference between this function and `get_args_4_rename_table_all()` is that this one does not include `"new_table_name"` in `args`. I don't see why it would be desirable to implement two functions that duplicate each other to such an extent.

**Review by eerovaher — COMMENTED** [review]
I have a few comments and suggestions, but also a few questions.

File: astroquery/gaia/core.py
Comment (mhsarmiento) [reply astroquery/gaia/core.py]: Change applied

File: astroquery/gaia/core.py
Comment (mhsarmiento) [reply astroquery/gaia/core.py]: Change applied. Thanks!

File: astroquery/gaia/core.py
Comment (mhsarmiento) [reply astroquery/gaia/core.py]: Change applied. Thanks!

File: astroquery/utils/tap/core.py
Comment (mhsarmiento) [reply astroquery/utils/tap/core.py]: Change applied. Thanks!

File: astroquery/utils/tap/core.py
Comment (mhsarmiento) [reply astroquery/utils/tap/core.py]: Suggested change applied. Thanks!

File: astroquery/gaia/core.py
Comment (mhsarmiento) [reply astroquery/gaia/core.py:932]: Suggested change applied. Thanks!

File: astroquery/utils/tap/tests/test_tap.py
Comment (mhsarmiento) [reply astroquery/utils/tap/tests/test_tap.py]: Suggested change applied. Thanks!

File: astroquery/utils/tap/tests/test_tap.py
Comment (mhsarmiento) [reply astroquery/utils/tap/tests/test_tap.py:976]: Not sure yet which type of exceptions are risen here. I have already open an issue on my side to improve this part of the code in future versions. Thank you very much for pointing this out.

File: docs/gaia/gaia.rst
Comment (mhsarmiento) [reply docs/gaia/gaia.rst]: Sorry for this. Was due to a problem during the rebase. I have restored the example.

File: docs/gaia/gaia.rst
Comment (mhsarmiento) [reply docs/gaia/gaia.rst]: Suggested change applied

File: docs/gaia/gaia.rst
Comment (mhsarmiento) [reply docs/gaia/gaia.rst]: Example updated.

Again thank you very much for your time and suggestions.

File: astroquery/gaia/core.py
Comment (mhsarmiento) [reply astroquery/gaia/core.py]: It is from the TAP+ protocol. By default certain type of output formats are returned in compress format

File: astroquery/gaia/core.py
Comment (mhsarmiento) [reply astroquery/gaia/core.py]: Suggested change applied. Thanks!

File: astroquery/utils/tap/conn/tapconn.py
Comment (mhsarmiento) [reply astroquery/utils/tap/conn/tapconn.py]: Suggested change applied. Thanks!

File: astroquery/utils/tap/conn/tapconn.py
Comment (mhsarmiento) [reply astroquery/utils/tap/conn/tapconn.py]: Suggested change applied. Thanks!

File: astroquery/utils/tap/core.py
Comment (mhsarmiento) [reply astroquery/utils/tap/core.py]: logic reviewed. Thank you for this suggestion.

File: astroquery/utils/tap/core.py
Comment (mhsarmiento) [reply astroquery/utils/tap/core.py]: Suggested change applied. Thanks!

File: astroquery/utils/tap/core.py
Comment (mhsarmiento) [reply astroquery/utils/tap/core.py]: Yes, that's correct! Suggested change applied. Thank you!

File: astroquery/utils/tap/core.py
Comment (mhsarmiento) [reply astroquery/utils/tap/core.py]: Thank you very much. Thank to you guys I'm learning and improving my knowledge of python. Thanks!

File: astroquery/utils/tap/core.py
Comment (mhsarmiento) [reply astroquery/utils/tap/core.py]: Suggested change applied

File: astroquery/gaia/core.py
Comment (eerovaher) [reply astroquery/gaia/core.py]: If it's a feature of the TAP+ protocol then it would be better to open a separate pull request for implementing this in the `TapPlus` class.

File: astroquery/utils/tap/tests/test_tap.py
Comment (bsipocz) [reply astroquery/utils/tap/tests/test_tap.py:976]: Having an issue open in the astroquery repo may be useful, too, so we see what has come up already.

File: astroquery/utils/tap/xmlparser/utils.py
Comment (bsipocz) [inline astroquery/utils/tap/xmlparser/utils.py:35]: We still support python3.7, and `BadGzipFile` has only been added into 3.8, so a workaround is needed here. I see two ways, but maybe there are others (I don't have a preference of which one you do):

1) - except on `OSError` as `BadGzipFile` inherits for that one. Based on the docs, `EOFError` or `zlib.error` could be options, too.

2) Add a custom exception based on the python version. Somewhat ugly, but it is the closest to what you're checking right now, and would allow a trivial cleanup once 3.7 support is dropped.
```
try:
    # We need python > 3.7 for this
    from gzip import BadGzipFile
except ImportError:
    class BadGzipFile(OSError):
        pass
```

File: astroquery/utils/tap/xmlparser/utils.py
Comment (bsipocz) [inline astroquery/utils/tap/xmlparser/utils.py:60]: Would this be a simple ascii table, or votable format? If the latter why not use `votable`?

File: astroquery/gaia/core.py
Comment (mhsarmiento) [reply astroquery/gaia/core.py]: Sorry, I'm not sure if I understand here what do you want me to do. TAP+ returning a compressed format is not something that has been implemented now. 
The problem here was that when a user gives a name, for example "myVotable.vot" for the file and he/she does not expect that the file is going to be provided in compressed format and then he/she tries to open the file, the user is going to get and error because the file is compressed. In order to avoid that, we were requested to rename the file with the correct extension if the user didn't provided that at first 'myVotable.vot.gz', so they don't get that error when opening the file.

File: astroquery/utils/tap/xmlparser/utils.py
Comment (mhsarmiento) [reply astroquery/utils/tap/xmlparser/utils.py:35]: Thank you very much. Working on this right now

File: astroquery/gaia/core.py
Comment (eerovaher) [reply astroquery/gaia/core.py]: If the `votable` and `fits` files are compressed only by the Gaia archive then this code belongs here. But from your previous answer I got the impression that those file types being compressed is required by the TAP+ standard, and if that is the case then this code needs to go to the `TapPlus` class so that all `astroquery` modules connecting to TAP+ services could benefit from it.

File: astroquery/utils/tap/xmlparser/utils.py
Comment (eerovaher) [reply astroquery/utils/tap/xmlparser/utils.py:60]: Can you provide a minimal code example?

Comment (mhsarmiento) [reply astroquery/utils/tap/xmlparser/utils.py:60]: Sorry, but with type of example? Do you want the JupiterNotebook with the error returned by astropy table if we don't do that conversion? Honestly I would not like to rollback the code just to show this but I have to, I will do it of course.

File: astroquery/gaia/core.py
Comment (mhsarmiento) [reply astroquery/gaia/core.py]: It required by TAP+. I can move it to TapPlus for the next implementation but we have dealing with this pull request since past June and I/we would really appreciate if we can move on with this. Specially if as you have suggested, we need to do a different pull request for the TAP+ modifications.  I will take this also into account from now on.

Comment (eerovaher) [reply astroquery/gaia/core.py]: This pull request would implement many improvements to `astroquery.gaia`, but at least a few of them seem to be independent from the others. The problem with such a large pull request is that it can only be merged all at once, so if any of the proposed improvements requires some more work then merging all the other improvements has to be delayed too. Opening a separate pull request for each independent improvement would preferable.

File: astroquery/utils/tap/xmlparser/utils.py
Comment (mhsarmiento) [reply astroquery/utils/tap/xmlparser/utils.py]: change applied. thank you again for the suggestion. very useful

File: astroquery/gaia/core.py
Comment (jespinosaar) [reply astroquery/gaia/core.py]: Dear @eerovaher , @mhsarmiento ,

I think there is a misunderstanding here. Considering previous comments, what you are proposing is that we include that warning as a check in the launch_job method from TAPPlus class, so the rest of our modules could benefit from it. Is that correct? And that this improvement should be included in a different pull request, am I wrong?

If that is the case, I propose this: 

- On Monday, I will extract this piece of code.
- I will create a new branch from astroquery main to include only this feature.
- After testing it and checking that it works for all ESA modules, I will create the pull request. As this is something simple (just a few lines), it should be merged fast.
- Once it is accepted and in astroquery main branch, we will rebase this PR branch to include it and remove the code from Gaia module.

What do you think?

Have a nice weekend!

Comment (eerovaher) [reply astroquery/gaia/core.py]: My understanding based on the answers above is that changing file name extensions of compressed formats would be useful for all TAP+ services, and in that case it should be implemented in the `TapPlus` class to avoid code repetition. But I am not familiar with the TAP+ standard so it is possible I have misunderstood something. If I am mistaken and this code is useful only for the Gaia archive then it should stay where it is.

If the code in question does belong in `TapPlus` then I would say it would have been better to open a separate pull request. But the code is already included here, so opening another pull request at this point might not be worth the effort. It is best to ask @keflavich, @bsipocz or @ceb8 what they think.

> As this is something simple (just a few lines), it should be merged fast.

I do not have write access to this repository, so I cannot make any promises about when this or any other pull request gets merged, but I would expect that smaller and simpler pull requests get merged faster.

File: astroquery/utils/tap/xmlparser/utils.py
Comment (mhsarmientoc) [reply astroquery/utils/tap/xmlparser/utils.py]: Ok. change applied. Thanks!

**Comment by bsipocz** [comment]
@mhsarmiento - I've rebased and removed the duplicated commits, and now it became quite visible that the first few commits were made by someone who didn't set their email address to be recognized by github. Please let me know if you would like to fix commits by `ciwuser <ciwuser@mail.com>`, and if yes what name and email address should I amend those commits. If we do that before merging it will get into the git history properly, after merge it's a bit more complicated to fix (and won't be really fixed on the commit level). Thanks!

**Review by bsipocz — APPROVED** [review]
This can be merged once my minor last question about contributor git history credits are answered.

**Comment by mhsarmiento** [comment]
> @mhsarmiento - I've rebased and removed the duplicated commits, and now it became quite visible that the first few commits were made by someone who didn't set their email address to be recognized by github. Please let me know if you would like to fix commits by `ciwuser <ciwuser@mail.com>`, and if yes what name and email address should I amend those commits. If we do that before merging it will get into the git history properly, after merge it's a bit more complicated to fix (and won't be really fixed on the commit level). Thanks!

Dear @bsipocz ,I have already fix the commits by `ciwuser` Thank you very for your work on this!

**Comment by bsipocz** [comment]
Thank you @mhsarmiento!

**Pull request merged**
